### ISSUE TITLE
Improve Transaction Tests Flags

### ISSUE THREAD

**Issue by JeremyRubin** [bitcoin/bitcoin#22865]
_Originally posted by @glozow in https://github.com/bitcoin/bitcoin/pull/21702#r684235896_

While trying to write transaction tests for CTV i've come across the pathological case of trying to specify flags.

Suppose I have a script that is `<X> CTV`. CTV is an OP_NOP4 takeover, so when DISCOURAGE_UPGRADABLE_NOPS is set but VERIFY_CTV flag is not, then it should fail. Therefore we need to mark DISCOURAGE_UPGRADABLE_NOPS as excluded. However, when we exclude that flag, then the flag maximality check wants it to fail, but the txn is consensus valid.

If we were to just update the trim/fill flag functions, then we would be overly restrictive because we would set things up to e.g. always exlcude or include a flag when in reality it depends on the specific flags being set or unset.

I think this can be addressed by adding 2 new fields to the test:

1) flags that should be always on
2) transactions that should be exempt from the maximality check

We can then duplicate some tests to test the combo of flags to check that NONE are excluded + the new NOP4 passes as well as the case where DISCOURAGE_NOPS is excluded (i.e., never set) that turning it back on (during the maximality check) is not a failure. 

Any thoughts on this approach?

**Comment by JeremyRubin** [comment]
Concrete example patch + txvalid vectors. Without the patch the first example would fail because of the DISCOURAGE rules kicking ON with CTV off. The second example would fail becuase when the DISCOURAGE rule is excluded, turning it off to check for failure fails when the CTV rule is on. These kinda sound like the same case, but they're a bit different...

```json
["Check that CTV is Processed with a Taproot Spend"],
[[["f90521604b56c392ffa17a01bcae5914b8cf7728cc6cec00d90838818cc5465f", 0, "1 0x20 0x24f5fe807bcee7774dc515f0b7ee8d6ae39eefd1b590264c52ff867e22c49419", 155000]],
"020000000001015f46c58c813808d900ec6ccc2877cfb81459aebc017aa1ff92c3564b602105f90000000000000000000ae80300000000000017a914ce0036ae7d49f06967dd92cc1ffff4a878c457f987d00700000000000017a91406e00c3b362e65e03507a2858d7b6499b668669887b80b00000000000017a9142ee42c65592c59b69bfefbd03781140c67e5232487a00f00000000000017a9146b3df16a1e6651d582ca6598900cb4f2d6c9dfb887881300000000000017a914877d55932d4f38b476d4db27e4efbe159ff0a07187701700000000000017a91441e9dc892e861d252d513d594ba833cd6bc8917087581b00000000000017a914b93075800c693dcc78b0553bf9d1cf879d76a02487401f00000000000017a914e9f0ea3a2cae0ad01114e2ec3502ef08bbc50af487282300000000000017a9149a645b5293bdf8be72cb9d1460bce7d64445cfad87102700000000000017a91451e5d6b2ee24ae128234c92245df3624620ea7d3870222209eb65498bfcd4eb90e61c2c5e323a9c16c8bfd8d53ba649915bcdb572099c12fb321c0b7e0105780185688d998a8f8438aa07637a5799755688ec80175cb26c0406e0200000000",
"NONE", "DEFAULT_CHECK_TEMPLATE_VERIFY_HASH"],
["Check that CTV upgradability works (taproot)"],
[[["f90521604b56c392ffa17a01bcae5914b8cf7728cc6cec00d90838818cc5465f", 0, "1 0x20 0x24f5fe807bcee7774dc515f0b7ee8d6ae39eefd1b590264c52ff867e22c49419", 155000]],
"020000000001015f46c58c813808d900ec6ccc2877cfb81459aebc017aa1ff92c3564b602105f90000000000000000000ae80300000000000017a914ce0036ae7d49f06967dd92cc1ffff4a878c457f987d00700000000000017a91406e00c3b362e65e03507a2858d7b6499b668669887b80b00000000000017a9142ee42c65592c59b69bfefbd03781140c67e5232487a00f00000000000017a9146b3df16a1e6651d582ca6598900cb4f2d6c9dfb887881300000000000017a914877d55932d4f38b476d4db27e4efbe159ff0a07187701700000000000017a91441e9dc892e861d252d513d594ba833cd6bc8917087581b00000000000017a914b93075800c693dcc78b0553bf9d1cf879d76a02487401f00000000000017a914e9f0ea3a2cae0ad01114e2ec3502ef08bbc50af487282300000000000017a9149a645b5293bdf8be72cb9d1460bce7d64445cfad87102700000000000017a91451e5d6b2ee24ae128234c92245df3624620ea7d3870222209eb65498bfcd4eb90e61c2c5e323a9c16c8bfd8d53ba649915bcdb572099c12fb321c0b7e0105780185688d998a8f8438aa07637a5799755688ec80175cb26c0406e0200000000",
"DISCOURAGE_UPGRADABLE_NOPS", "NONE", true],
```


```diff
diff --git a/src/test/transaction_tests.cpp b/src/test/transaction_tests.cpp
index df2071d2b..717e30276 100644
--- a/src/test/transaction_tests.cpp
+++ b/src/test/transaction_tests.cpp
@@ -198,11 +198,19 @@ BOOST_AUTO_TEST_CASE(tx_valid)
         std::string strTest = test.write();
         if (test[0].isArray())
         {
-            if (test.size() != 3 || !test[1].isStr() || !test[2].isStr())
+            const size_t size = test.size();
+            const bool default_args = size == 3;
+            const bool has_skip_exclude_one = size == 5;
+            const bool has_extra_flags = size == 4;
+            const bool size_correct = default_args || has_extra_flags || has_skip_exclude_one;
+            const bool extra_flags_correct = !has_extra_flags || test[3].isStr();
+            const bool skip_exclude_one_correct = !has_skip_exclude_one  || test[4].isBool();
+            if (!size_correct || !test[1].isStr() || !test[2].isStr() || !extra_flags_correct || !skip_exclude_one_correct)
             {
                 BOOST_ERROR("Bad test: " << strTest);
                 continue;
             }
+            const bool skip_exclude_one = has_skip_exclude_one? test[4].get_bool() : false;
 
             std::map<COutPoint, CScript> mapprevOutScriptPubKeys;
             std::map<COutPoint, int64_t> mapprevOutValues;
@@ -243,33 +251,36 @@ BOOST_AUTO_TEST_CASE(tx_valid)
 
             PrecomputedTransactionData txdata(tx);
             unsigned int verify_flags = ParseScriptFlags(test[2].get_str());
+            unsigned int extra_verify_flags = has_extra_flags? ParseScriptFlags(test[3].get_str()) : 0;
 
             // Check that the test gives a valid combination of flags (otherwise VerifyScript will throw). Don't edit the flags.
             if (~verify_flags != FillFlags(~verify_flags)) {
                 BOOST_ERROR("Bad test flags: " << strTest);
             }
 
-            BOOST_CHECK_MESSAGE(CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, ~verify_flags, txdata, strTest, /* expect_valid */ true),
+            BOOST_CHECK_MESSAGE(CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, extra_verify_flags|~verify_flags, txdata, strTest, /* expect_valid */ true),
                                 "Tx unexpectedly failed: " << strTest);
 
             // Backwards compatibility of script verification flags: Removing any flag(s) should not invalidate a valid transaction
             for (const auto& [name, flag] : mapFlagNames) {
                 // Removing individual flags
-                unsigned int flags = TrimFlags(~(verify_flags | flag));
+                unsigned int flags = TrimFlags(extra_verify_flags | ~(verify_flags | flag));
                 if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, flags, txdata, strTest, /* expect_valid */ true)) {
                     BOOST_ERROR("Tx unexpectedly failed with flag " << name << " unset: " << strTest);
                 }
                 // Removing random combinations of flags
-                flags = TrimFlags(~(verify_flags | (unsigned int)InsecureRandBits(mapFlagNames.size())));
+                flags = TrimFlags(extra_verify_flags | ~(verify_flags | (unsigned int)InsecureRandBits(mapFlagNames.size())));
                 if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, flags, txdata, strTest, /* expect_valid */ true)) {
                     BOOST_ERROR("Tx unexpectedly failed with random flags " << ToString(flags) << ": " << strTest);
                 }
             }
 
             // Check that flags are maximal: transaction should fail if any unset flags are set.
-            for (auto flags_excluding_one : ExcludeIndividualFlags(verify_flags)) {
-                if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, ~flags_excluding_one, txdata, strTest, /* expect_valid */ false)) {
-                    BOOST_ERROR("Too many flags unset: " << strTest);
+            if (!skip_exclude_one) {
+                for (auto flags_excluding_one : ExcludeIndividualFlags(verify_flags)) {
+                    if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, ~flags_excluding_one, txdata, strTest, /* expect_valid */ false)) {
+                        BOOST_ERROR("Too many flags unset: " << strTest);
+                    }
                 }
             }
         }
```

**Comment by glozow** [comment]
Just to make sure I understand you correctly, please correct me if this is inaccurate:

Your test is a valid OP_CTV transaction. We expect the following results when we pass in these combinations of script verification flags:
```
DISCOURAGE_NOPS      CHECKTEMPLATEVERIFY      expected result
      on                 on                 fail because using NOP4
      on                 off                fail because using NOP4
      off                on                 success, this result is what we're interested in
      off                off                success because it should be a soft fork
```

The transaction_tests framework expects the set of excluded flags to be minimal, i.e. removing any more flags should be fine, but adding any flags should make it fail. Flags are specified as _excluded_ and the default passed in is everything in `STANDARD_SCRIPT_VERIFY_FLAGS`. If your test is failing the flag maximality check when you specify exclude=`DISCOURAGE_UPGRADEABLE_NOPS`, that means either:

- It passed in an extra flag and it didn't fail. This means your test needs to be more specific.
- It removed an extra flag and it failed. Note that if it removed CTV and it failed, this means CTV isn't a soft fork. There are other possibilities as well.

**Comment by JeremyRubin** [comment]
You're missing what's going on. DISCOURAGE_UPGRADABLE_NOPS is a non consensus flag.

You should be able to e.g. test the following script spend:

`<H> CTV`

in the correct txn for H.

- When VERIFY_CTV is OFF and DISCOURAGE is ON it will fail.
- When VERIFY_CTV is OFF and DISCOURAGE is OFF it will pass.
- When VERIFY_CTV is ON and DISCOURAGE is ON it will pass.
- When VERIFY_CTV is ON and DISCOURAGE is OFF it will pass.

Therefore we must either set either VERIFY_CTV is always on, or DISCOURAGE is always off and skip the DISCOURAGE minimal test.

Now consider the spend:

`1 CTV`

- When VERIFY_CTV is OFF and DISCOURAGE is ON it will fail.
- When VERIFY_CTV is OFF and DISCOURAGE is OFF it will pass.
- When VERIFY_CTV is ON and DISCOURAGE is ON it will fail.
- When VERIFY_CTV is ON and DISCOURAGE is OFF it will pass.

Therefore, if we would like to test this, we need to specify that just DISCOURAGE is OFF (covered by existing behavior). 


This is also why we cannot use FillFlags or TrimFlags afaict, since we don't know if the transaction is intended to succeed or fail (<H> CTV or 1 CTV) so we can't infer the correct fill.
------


We cannot simply change the line back to the form that does not reture discourage upgradable because we need to discourage using such features to protect old mining nodes during a soft fork upgrade in the future.

This was botched with CSV; if we wanted to add a new CSV type that uses the disabled flag, we would have to wait to age out old mining nodes for it to be safe for them to continue to mine on the network. Hence my patch to fix it, so we get a head start on that arduous process.

**Comment by glozow** [comment]
No, I think we have a miscommunication here. When I say a flag is a soft fork, I mean that applying the flag can only restrict the space of acceptable scripts - see #10699

> in the correct txn for H.
    When VERIFY_CTV is OFF and DISCOURAGE is ON it will fail.
    When VERIFY_CTV is ON and DISCOURAGE is ON it will pass.

In the example you're giving here, VERIFY_CTV is not a soft fork, since applying it to this transaction causes it to go from invalid to valid. Thus, I don't think it is the right approach. If you want templates to be upgradeable, you can version them and discourage greater versions, allow an extra field and discourage using it, etc.

> We cannot simply change the line back to the form that does not reture discourage upgradable because we need to discourage using such features to protect old mining nodes during a soft fork upgrade in the future.

I agree with protecting old nodes that haven't upgraded in soft forks, but I still don't understand why you're doing it this way. IIUC since #5000 they should be discouraging NOP4 in policy already.

**Comment by JeremyRubin** [comment]
Hard forks / soft fork only applies in that way to consensus flags, not standardness. DISCOURAGE_UPGRADABLE_NOPS is never used in consensus.

Taproot does something sorta similar currently, albeit in Input standardness rather than interpreter although the distinction is immaterial.

```c++
    // Check for non-standard pay-to-script-hash in inputs
    const bool taproot_active = DeploymentActiveAfter(m_active_chainstate.m_chain.Tip(), args.m_chainparams.GetConsensus(), Consensus::DEPLOYMENT_TAPROOT);
    if (fRequireStandard && !AreInputsStandard(tx, m_view, taproot_active)) {
        return state.Invalid(TxValidationResult::TX_INPUTS_NOT_STANDARD, "bad-txns-nonstandard-inputs");
    }
```

Therefore when taproot_active goes from false -> true, a transaction goes from invalid to valid. It's still a soft fork.


Part of the reason why we need the discourage in the case that SCRIPT_VERIFY_DEFAULT_CHECK_TEMPLATE_VERIFY_HASH is not set is (if it helps to reason about it) what if CTV never gets activated? We should perfectly preserve the old NOP semantics of discouraging it.

**Comment by JeremyRubin** [comment]
Something else that might help to see this more clearly:

The (current BIP119) code for CTV & NOPs:
```c++
               case OP_CHECKTEMPLATEVERIFY:
                {
                    // if flags not enabled; treat as a NOP4
                    if (!(flags & SCRIPT_VERIFY_DEFAULT_CHECK_TEMPLATE_VERIFY_HASH)) {
                        if (flags & SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS)
                            return set_error(serror, SCRIPT_ERR_DISCOURAGE_UPGRADABLE_NOPS);
                        break;
                    }

                    if (stack.size() < 1)
                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);

                    // If the argument was not 32 bytes, treat as OP_NOP4:
                    switch (stack.back().size()) {
                        case 32:
                            if (!checker.CheckDefaultCheckTemplateVerifyHash(stack.back())) {
                                return set_error(serror, SCRIPT_ERR_TEMPLATE_MISMATCH);
                            }
                            break;
                        default:
                            // future upgrade can add semantics for this opcode with different length args
                            // so discourage use when applicable
                            if (flags & SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS) {
                                return set_error(serror, SCRIPT_ERR_DISCOURAGE_UPGRADABLE_NOPS);
                            }
                    }
                }
                break;

                case OP_NOP1: case OP_NOP5:
                case OP_NOP6: case OP_NOP7: case OP_NOP8: case OP_NOP9: case OP_NOP10:
                {
                    if (flags & SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS)
                        return set_error(serror, SCRIPT_ERR_DISCOURAGE_UPGRADABLE_NOPS);
                }
                break;
```
The pruned code if SCRIPT_VERIFY_CHECK_TEMPLATE_VERIFY_HASH is always set:
```c++
               case OP_CHECKTEMPLATEVERIFY:
                {
                    if (stack.size() < 1)
                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);

                    // If the argument was not 32 bytes, treat as OP_NOP4:
                    switch (stack.back().size()) {
                        case 32:
                            if (!checker.CheckDefaultCheckTemplateVerifyHash(stack.back())) {
                                return set_error(serror, SCRIPT_ERR_TEMPLATE_MISMATCH);
                            }
                            break;
                        default:
                            // future upgrade can add semantics for this opcode with different length args
                            // so discourage use when applicable
                            if (flags & SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS) {
                                return set_error(serror, SCRIPT_ERR_DISCOURAGE_UPGRADABLE_NOPS);
                            }
                    }
                }
                break;
```
The (pruned) code if CTV is never set.
```c++
               case OP_CHECKTEMPLATEVERIFY:
                {
                    // if flags not enabled; treat as a NOP4
                    if (flags & SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS)
                        return set_error(serror, SCRIPT_ERR_DISCOURAGE_UPGRADABLE_NOPS);
                }
                break;

                case OP_NOP1: case OP_NOP5:
                case OP_NOP6: case OP_NOP7: case OP_NOP8: case OP_NOP9: case OP_NOP10:
                {
                    if (flags & SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS)
                        return set_error(serror, SCRIPT_ERR_DISCOURAGE_UPGRADABLE_NOPS);
                }
                break;
```
Which is the same as the current code:

```c++
                case OP_NOP1: case OP_NOP4: case OP_NOP5:
                case OP_NOP6: case OP_NOP7: case OP_NOP8: case OP_NOP9: case OP_NOP10:
                {
                    if (flags & SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS)
                        return set_error(serror, SCRIPT_ERR_DISCOURAGE_UPGRADABLE_NOPS);
                }
                break;
```
The pruned code if SCRIPT_VERIFY_CHECK_TEMPLATE_VERIFY_HASH is always set and the argument is known to be 32 bytes:
```c++
               case OP_CHECKTEMPLATEVERIFY:
                {
                    if (stack.size() < 1)
                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
                    if (!checker.CheckDefaultCheckTemplateVerifyHash(stack.back())) {
                        return set_error(serror, SCRIPT_ERR_TEMPLATE_MISMATCH);
                    }
                }
                break;

```
The pruned code if SCRIPT_VERIFY_CHECK_TEMPLATE_VERIFY_HASH is always set and the argument is known to not be 32 bytes:

```c++
               case OP_CHECKTEMPLATEVERIFY:
                {
                    if (stack.size() < 1)
                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
                    if (flags & SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS) {
                        return set_error(serror, SCRIPT_ERR_DISCOURAGE_UPGRADABLE_NOPS);
                    }
                 }                }
                break;
```

The list below (I hope?) makes it more plain to see:

- When CTV is not active, the semantics are entirely identical to CTV being NOP4.
  - During Consensus, Always Passes
  - During Standardness, Always Fails (DISCOURAGE_UPGRADABLE_NOPS)
- When CTV is active, there must always be at least 1 element on the stack (according to @roconnor-blockstream it is a very good thing if opcodes never take a variable # of things off the stack, so we do not allow CTV with an empty stack to have a meaning).
  - During Consensus, Passes if >=1 element on stack
  - During Standardness, Passes if >=1 element on stack
- When CTV is active, and the argument is 32 bytes, we apply CTV rules
  - During Consensus, Passes if CTV Hash Matches TXN
  - During Standardness, Passes if CTV Hash Matches TXN
- When CTV is active, and the argument is not 32 bytes, we continue to treat as a NOP (I could probably clarify the situation a bit better here by making a new flag -- DISCOURAGE_UPGRADABLE_TEMPLATE_PROGRAM, but the first DISCOURAGE_UPGRADABLE_NOP check still causes this problem by itself).
  - During Consensus, always passes
  - During Standardness, always fails (DISCOURAGE_UPGRADABLE_NOPS)

Therefore we're only ever adding cases where something can fail in consensus, and standardness is more strict that consensus.

**Comment by glozow** [comment]
> Taproot does something sorta similar currently, albeit in Input standardness rather than interpreter although the distinction is immaterial.
Therefore when taproot_active goes from false -> true, a transaction goes from invalid to valid. It's still a soft fork.

To clarify, I'm talking about soft forks wrt script verification flags as inputs to the interpreter, not about our policy rules. I don't think the distinction is immaterial. Yes, I am aware that when taproot activates we will no longer discourage taproot spends in policy, while consensus rules will tighten, since it is a soft fork.

What I am trying to say is: [a property we desire in the script verification flags](https://github.com/bitcoin/bitcoin/blob/e7c6ed605ca4991856603971f1306dd484dc2f75/src/script/interpreter.h#L37-L41) is that the addition of another flag strictly reduces the space of acceptable scripts. This property is what is being tested by the maximality checks in transaction_tests; they fail because `SCRIPT_VERIFY_CHECK_TEMPLATE_VERIFY_HASH` does not have this property. My suggestion is rather than bypassing this check in transaction_tests, implement CTV extensibility and maintain this property by adding another script verification flag. A method is outlined [here](https://github.com/bitcoin/bitcoin/pull/10699#issue-128042516).

> When CTV is active, and the argument is not 32 bytes, we continue to treat as a NOP (I could probably clarify the situation a bit better here by making a new flag -- DISCOURAGE_UPGRADABLE_TEMPLATE_PROGRAM, but the first DISCOURAGE_UPGRADABLE_NOP check still causes this problem by itself).

I believe the approach of having a `SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_TEMPLATE_PROGRAM` would be better. Again, I don't think `SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_NOPS` should be in the logic that handles the now-repurposed NOP4.

**Comment by JeremyRubin** [comment]
Soft fork/hard fork refers strictly to consensus, not to policy. DISCOURAGE_UPGRADABLE_NOPS is policy.

I don't think we can -- while preserving the (A|B) is subset A -- safely do any OP_NOP upgrade since we need some flag which is DISCOURAGE_UNACTIVE_UPGRADED_BLAH to discourage using a thing before the upgrade is active but where it is still defined. This will be a problem for Taproot, CTV, APO, etc. So it seems either we must break this property of (A|B) is subset A while doing an upgrade, or we don't do any upgrades?

**Comment by JeremyRubin** [comment]
Soft fork/hard fork refers strictly to consensus, not to policy. DISCOURAGE_UPGRADABLE_NOPS is policy.

I don't think we can -- while preserving the (A|B) is subset A -- safely do any OP_NOP upgrade since we need some flag which is DISCOURAGE_UNACTIVE_UPGRADED_BLAH to discourage using a thing before the upgrade is active but where it is still defined. This will be a problem for Taproot, CTV, APO, etc. So it seems either we must break this property of (A|B) is subset A while doing an upgrade, or we don't do any upgrades?

In other words, we fundamentally need some sort of IS_EXECUTING_IN_CONSENSUS/IS_EXECUTING_IN_POLICY flag to distinguish if using a to-be activated feature before it is active should be a failure or not. 

(sorry for the deleted comment, accident)

**Comment by sipa** [comment]
I don't see why this property needs to be broken. To the best of my knowledge, every script validation flag today, is purely restrictive (in the sense that the set of valid scripts under A|B is a subset of the scripts valid under A). It is true that such a property is not strictly required for anything but consensus rules (which the various DISCOURAGE ones aren't), but it is extremely valuable for testing I think.

I haven't followed the discussion, feel free to explain what I'm missing, or where I can find it.

**Comment by sipa** [comment]
@JeremyRubin Thanks for elaborating.

Historically I think we used a different mechanism than Taproot used, where as soon as a new opcode was defined, we'd unconditionally remove DISCOURAGE semantics from it, and just always fully enforce it (correctly), even pre-activation, for mempool transactions.

It hadn't occurred to me that this approach we're taking with Taproot (where even valid taproot spends remain non-standard until activation) is in fact only possible because it's done through non-script IsStandard() rules, and doing the same through DISCOURAGE flags would violate the restricting-only property.

However, I do think it's better to take that path, and treat future opcode spends as non-standard until activation. At the very least, it means a more predictable relay policy, new nodes switch policy in lockstep, rather than following adoption patterns.

That will mean indeed that adding FLAG_NEW_OPCODE to FLAG_DISCOURAGE is not a pure restriction, but that sounds like a reasonable price to pay.

**Comment by JeremyRubin** [comment]
The patch https://github.com/bitcoin/bitcoin/pull/22876 makes an attempt at at least preserving testability until we know it is fully active by allowing to test the same transaction with flags as below:

```
// tests with DISCOURAGE_UPGRADABLE_NOPS must be unset, no flags required to be set, and the maximality check skipped
...'DISCOURAGE_UPGRADABLE_NOPS', 'NONE', true]
// tests with all flags ok, SOME_OPCODE_VERIFY flags required to be set, and the maximality check enforced
...'NONE', 'SOME_OPCODE_VERIFY']
```

I think it's less clean than ideal, but allows cover both cases of soft fork succeeds and soft fork fails.

**Comment by ajtowns** [comment]
I think there's four interpretations we want; one pre CTV activation:

 - `x OP_NOP4` -- consensus valid, relay invalid, doesn't matter what x might be

and three post CTV activation:

 - `goodhash OP_CTV` -- fully specced and enforced per BIP119, consensus valid, relay valid
 - `badhash OP_CTV` -- hash doesn't match tx, consensus invalid
 - `upgrade OP_CTV` -- part of upgradeability, consensus valid, relay invalid

The specific case causing a conflict is that you want `goodhash OP_CTV` to go from "relay invalid" prior to CTV activating to "relay valid" after it activates which is removing a restriction, but indicating CTV is active is adding a flag, and it would be nice if that made things more restrictive rather than less restrictive.

Perhaps ideally we'd say `DISCOURAGE_CTV` prior to it being active?

Maybe it would be better to explicitly add a flag like `DISCOURAGE_CTV` or `DISCOURAGE_TAPROOT` for preventing relay  when the softfork is not active; then remove that flag once the fork is active and buried? That would preserve the "flags only restrict things" and still get the full variety of behaviours, and only mean adding a two flags instead of one temporarily?

If we wanted to ensure new nodes didn't relay CTV transactions before they'd caught up to CTV activation even after `DISCOURAGE_CTV` was removed from the codebase,  we could keep that behaviour by making standardness logic be "don't relay **anything** until your tip sees all the expected soft forks as active".

**Comment by JeremyRubin** [comment]
Would this approach work for anyprevout key types as well?

**Comment by JeremyRubin** [comment]
I generally philosophically feel a bit funky about taking on more code complexity in consensus logic, where extending the testing logic complexity would do. But if it's a clear cut approach that works generally I might feel different.

**Comment by ajtowns** [comment]
> Would this approach work for anyprevout key types as well?

I think the idea is that current behaviour would be just:

| key | sig | flags=nothing | DISCOURAGE |
|---|---|---|---|
| apo + future | * | valid | invalid |

But with APO support that would become:

| key | sig | flags=nothing | DISCOURAGE | APO | DISCOURAGE_APO |
|---|---|---|---|---|---|
| apo | empty/valid | valid | valid | valid | invalid |
| apo | badsig | valid | valid | invalid | invalid |
| future | * | valid | invalid | valid | valid |

(And when you combine flags, the result is valid iff all the individual results were valid)

And you'd change `PolicyScriptChecks` to something like:

```c++
    const unsigned int discourageflags = DeploymentActiveAfter(Tip(), ANYPREVOUT) ? 0 : DISCOURAGE_APO;
    const unsigned int scriptVerifyFlags = STANDARD_SCRIPT_VERIFY_FLAGS | discourageflags;
```

(with `VERIFY_APO` included in `STANDARD_SCRIPT_VERIFY_FLAGS`)

**Comment by JeremyRubin** [comment]
I'd need to stew on this more and maybe see a PoC working. My brain hurts!

The issue I see is that we need to not only ensure the "invariants" hold for both tx_valid and tx_invalid checks, and it seems hard to ensure that these flags are valid for all the test patterns and combinations we try.

For tx_valid, we need to have that any additional flag set makes it invalid,right?

And for tx_invalid, we need to have that any addtional flag unset makes it valid, right?

So if we had an invalid tx, it seems that *both* the APO and DISCOURAGE_APO flag being set, unsetting just one still yields and invalid transaction...

**Comment by ajtowns** [comment]
Hmm, fair: in my head I was thinking you'd set one of APO or DISCOURAGE_APO not both, but then when I looked at the code figured you'd want to have APO in STANDARD_SCRIPT_VERIFY_FLAGS, which would mean setting both. Having `DISCOURAGE_VALID_APO` instead avoids that inconsistency.

```c++
    } else if ((pubkey.size() == 1 || pubkey.size() == 33) && pubkey[0] == 0x01) {
        if (flags & SCRIPT_VERIFY_ANYPREVOUT) != 0) {
            // check valid APO
            if (not valid APO) {
                 return false; // serror is set
            }
            if (flags & SCRIPT_VERIFY_DISCOURAGE_VALIDANYPREVOUT) != 0) {
                 return set_error(serror, TAPROOT_DISCOURAGE_VALIDANYPREVOUT);
            }
        }
    } else { ... }
```

> For tx_valid, we need to have that any additional flag set makes it invalid,right?

Hmm. Yeah, the tests are:

* tx_valid:
  * we've specified the maximal set of flags, adding any will make it invalid 
  * removing any flags will not make it invalid
* tx_invalid:
  * removing any flags will make it valid
  * adding any flags will not make it valid

With the caveat that the flags specified for tx_valid are inverted -- you specify the flags that would make it invalid, which is usually NONE.

> So if we had an invalid tx, it seems that _both_ the APO and DISCOURAGE_APO flag being set, unsetting just one still yields and invalid transaction...

This just means you test it by having the same tx specified twice in tx_invalid, once with just the APO flag and once with just the DISCOURAGE_APO flag? I think doing DISCOURAGE_VALID_APO solves this though, since then you just have one test with both flags, and removing either makes it valid.

**Comment by JeremyRubin** [comment]
So we need the following:
```
DISCOURAGE_APO
DISCOURAGE_VALID_APO
DISCOURAGE_NEW_KEYTYPE
VALIDATE_APO
```

and we do 2 copies of each transaction in tx_invalid? (Another option here would be to allow for lists of lists of flag combos, and re-run each test across a couple different rules.)

And after activation, we bury it to be:
```
DISCOURAGE_NEW_KEYTYPE
VALIDATE_APO
```

Overall this feels like a lot of complexity for reviewers on the consensus code v.s. just making the testing logic handle it.

Do you think there are other ways we could clean up our testing strategies / flag passing strategy to make things more clean?

**Comment by ajtowns** [comment]
We currently have `DISCOURAGE_UPGRADABLE_PUBKEYTYPE`.

When adding APO, we'd add `SCRIPT_VERIFY_APO` and `SCRIPT_VERIFY_DISCOURAGE_VALID_APO`. 

Tests of valid script path using APO would specify DISCOURAGE_VALID_APO in the tx_valid VERIFY_FLAGS, rather than NONE.

Tests of invalid script path spends using APO would specify `VERIFY_APO` as the minimal flag to make them invalid, rather than `DISCOURAGE_UPGRADABLE_PUBKEYTYPE` (which no longer makes them invalid).

When burying APO, we'd drop `SCRIPT_VERIFY_DISCOURAGE_VALID_APO` -- no longer specifying it for the tests, removing the `if () { return set_error(discouraged); }`, and no longer setting the flag for mempool entry when APO is not yet active.


That is, there's no DISCOURAGE_APO flag, just DISCOURAGE_VALID_APO, and no need to duplicate txs for testing. At least, I think that all works.

**Comment by Osusushue** [comment]
E

**Comment by ajtowns** [comment]
Here's what I've ended up doing in https://github.com/bitcoin-inquisition/bitcoin/pull/4/files:

 * add `SCRIPT_VERIFY_ANYPREVOUT` and `SCRIPT_VERIFY_DISCOURAGE_ANYPREVOUT`
 * remove APO pubkeys (ie `0x01` and `0x01[32 bytes]`) from `SCRIPT_VERIFY_DISCOURAGE_UPGRADEABLE_PUBKEYTYPE`
 * if `DISCOURAGE_ANYPREVOUT` is set, invalid if an APO pubkey is attempted to be used
 * if `ANYPREVOUT` is set, invalidate if an APO pubkey is used and the signature doesn't comply with BIP 118
 * set `DISCOURAGE_ANYPREVOUT` in `MemPoolAccept::PolicyScriptChecks` when APO is not activated

After activation, drop `DISCOURAGE_ANYPREVOUT` and remove code paths that set it, or that were only executed if it was set.

(I ended up having `DISCOURAGE_APO` instead of `DISCOURAGE_VALID_APO` rather than the other way around; but either way, there's only one, not both)

**Comment by ajtowns** [comment]
Possibly relevant https://github.com/bitcoin/bitcoin/pull/5253#issuecomment-67148736

**Comment by JeremyRubin** [comment]
good find!
